typing/typecore.ml

```diff
@@ -975,11 +975,18 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
   end
 
 let solve_Ppat_array ~refine loc env expected_ty =
-  let ty_elt = newgenvar() in
   let expected_ty = generic_instance expected_ty in
-  unify_pat_types_refine ~refine
-    loc env (Predef.type_array ty_elt) expected_ty;
-  ty_elt
+  if is_principal expected_ty &&
+     match get_desc (expand_head !!env expected_ty) with
+     | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+     | _ -> false
+  then
+    Predef.type_float, Predef.type_floatarray
```

Comment:
I don't think this needs to return the `type_floatarray` here. The existing treatment of the pattern type in `type_pat_aux` should work.

Reply:
Thanks, fixed.